x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationUtils.java

```diff
@@ -153,4 +198,182 @@ public static void switchUserBasedOnActionOriginAndExecute(
+        TransportRequest request,
+        String childAction,
+        Version version,
+        ClusterState clusterState
+
+    ) {
+        if (RemoteConnectionManager.resolveRemoteClusterAlias(connection).isPresent()) {
+            // We can only pre-authorize actions targeting a node which belongs to the same cluster.
+            if (threadContext.getHeader(ParentActionAuthorization.THREAD_CONTEXT_KEY) != null) {
+                if (logger.isDebugEnabled()) {
+                    logger.debug(
+                        "attempted to send the request ["
+                            + childAction
+                            + "] to the remote cluster ["
+                            + RemoteConnectionManager.resolveRemoteClusterAlias(connection).get()
+                            + "] with the pre-authorization header ["
+                            + threadContext.getHeader(ParentActionAuthorization.THREAD_CONTEXT_KEY)
+                            + "]"
+                    );
+                }
+                throw new AssertionError("pre-authorization not expected to be set for remote cluster requests");
+            }
+            return;
+        }
+
+        final AuthorizationEngine.AuthorizationContext parentContext = readParentAuthorizationContext(threadContext);
+        if (parentContext == null) {
+            return;
+        }
+
+        final IndicesAccessControl indicesAccessControl = parentContext.getIndicesAccessControl();
+        if (indicesAccessControl == null) {
+            // This can happen if the parent request was authorized by index name only - e.g. bulk request
+            // A missing IAC is not an error, but it means we can't safely tie authz of the child action to the parent authz
+            return;
+        }
+
+        // Just a sanity check. If we ended up here, the authz should have been granted.
+        if (indicesAccessControl.isGranted() == false) {
+            return;
+        }
+
+        final IndicesRequest indicesRequest;
+        if (request instanceof IndicesRequest) {
+            indicesRequest = (IndicesRequest) request;
+        } else {
+            // We can only pre-authorize indices request.
+            return;
+        }
+
+        final Role role = RBACEngine.maybeGetRBACEngineRole(parentContext.getAuthorizationInfo());
+        if (role == null) {
+            // If role is null, it means a custom authorization engine is in use, hence we cannot do the optimization here.
+            return;
+        }
+
+        // We can't safely pre-authorize actions if DLS or FLS is configured without passing IAC as well with the authorization result.
+        // For simplicity, we only pre-authorize actions when FLS and DLS are not configured, otherwise we would have to compute and send
+        // indices access control as well, which we want to avoid with this optimization.
+        if (role.hasFieldOrDocumentLevelSecurity()) {
+            return;
+        }
```

Comment:
The role DLS permission check is redundant. It's enough to check for DLS on the indices permission.

Reply:
I completely agree, but I'd prefer to keep it as a sanity check.